soroban-sdk/src/crypto.rs

```diff
@@ -251,3 +253,381 @@ impl CryptoHazmat {
+
+            pub fn as_val(&self) -> &Val {
+                self.0.as_val()
+            }
+
+            pub fn to_val(&self) -> Val {
+                self.0.to_val()
+            }
+
+            pub fn as_object(&self) -> &BytesObject {
+                self.0.as_object()
+            }
+
+            pub fn to_object(&self) -> BytesObject {
+                self.0.to_object()
+            }
+        }
+
+        impl IntoVal<Env, Val> for $elem {
+            fn into_val(&self, e: &Env) -> Val {
+                self.0.into_val(e)
+            }
+        }
+
+        impl TryFromVal<Env, Val> for $elem {
+            type Error = ConversionError;
+
+            fn try_from_val(env: &Env, val: &Val) -> Result<Self, Self::Error> {
+                let bytes = <BytesN<$size>>::try_from_val(env, val)?;
+                Ok($elem(bytes))
+            }
+        }
+
+        impl IntoVal<Env, BytesN<$size>> for $elem {
+            fn into_val(&self, _e: &Env) -> BytesN<$size> {
+                self.0.clone()
+            }
+        }
+
+        impl From<$elem> for Bytes {
+            fn from(v: $elem) -> Self {
+                v.0.into()
+            }
+        }
+
+        impl From<$elem> for BytesN<$size> {
+            fn from(v: $elem) -> Self {
+                v.0
+            }
+        }
+
+        impl Into<[u8; $size]> for $elem {
+            fn into(self) -> [u8; $size] {
+                self.0.into()
+            }
+        }
+
+        impl Eq for $elem {}
+
+        impl PartialEq for $elem {
+            fn eq(&self, other: &Self) -> bool {
+                self.0.partial_cmp(other.as_bytes()) == Some(Ordering::Equal)
+            }
+        }
+
+        impl Debug for $elem {
+            fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+                write!(f, "{}({:?})", $name, self.to_array())?;
+                Ok(())
+            }
+        }
+    };
+}
+
+impl_bls_elements_bytes_repr!(G1Affine, 96, "G1Affine");
+impl_bls_elements_bytes_repr!(G2Affine, 192, "G2Affine");
+impl_bls_elements_bytes_repr!(Fp, 48, "Fp");
+impl_bls_elements_bytes_repr!(Fp2, 96, "Fp2");
+
+impl Bls12_381 {
+    pub(crate) fn new(env: &Env) -> Bls12_381 {
+        Bls12_381 { env: env.clone() }
+    }
+
+    pub fn env(&self) -> &Env {
+        &self.env
+    }
+
+    // g1
+
+    /// Checks if a point `p` in G1 is in the correct subgroup.
+    pub fn g1_is_in_subgroup(&self, p: &G1Affine) -> bool {
+        let env = self.env();
+        let res = internal::Env::bls12_381_check_g1_is_in_subgroup(env, p.to_object())
+            .unwrap_infallible();
+        res.into()
+    }
+
+    /// Adds two points `p0` and `p1` in G1.
+    pub fn g1_add(&self, p0: &G1Affine, p1: &G1Affine) -> G1Affine {
+        let env = self.env();
+        let bin = internal::Env::bls12_381_g1_add(env, p0.to_object(), p1.to_object())
+            .unwrap_infallible();
+        G1Affine::from_bytes(bin.try_into_val(env).unwrap_optimized())
```

Comment:
nit: Could try_into_val.unwrap be just into_val?

Reply:
done